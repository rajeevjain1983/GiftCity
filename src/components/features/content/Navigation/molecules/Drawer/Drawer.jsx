// 9fbef606107a605d69c0edbcd8029e5d
import React from "react";
import withStyles from "../../../../../common/hoc/withStyle";
import { breakpoints } from "../../../../../../../styles/themes/mediaQuery";
import {
  showOverlay,
  closeOverlay,
  enableBodyScroll,
  disableBodyScroll,
  removeBodyScrollLocks,
  isMobileWeb,
} from "../../../../../../utils";
import { Row } from "../../../../../common/atoms";
import style from "./Drawer.style";

/**
 * Determines if Drawer is not required on all view ports
 * @param {Bool} small
 * @param {Bool} medium
 * @param {Bool} large
 */
const isDrawerNotRequiredOnAllViewports = (small, medium, large) => {
  return !(small && medium && large);
};

const hideOnViewport = (viewport) => {
  return `${!viewport.small ? "display-small-none" : ""} ${
    !viewport.medium ? "display-medium-none" : ""
  } ${!viewport.large ? "display-large-none" : ""}`;
};

const showOnViewport = (viewport) => {
  return `${viewport.small ? "display-small-none" : ""} ${
    viewport.medium ? "display-medium-none" : ""
  } ${viewport.large ? "display-large-none" : ""}`;
};

const renderDrawerFooter = (hideNavigationFooter, drawerFooter) => {
  const Footer = drawerFooter;
  let classToHide = "";
  if (hideNavigationFooter) {
    classToHide = "is-hidden";
  }
  return (
    drawerFooter && (
      <Footer
        className={`navigation-footer ${classToHide}`}
        isNavigationFooter
      />
    )
  );
};

class Drawer extends React.Component {
  constructor(props) {
    super(props);
    this.darkOverlayClass = "dark-overlay";
  }

  componentDidMount() {
    if (this.drawerRef && isMobileWeb()) {
      this.drawerRef.addEventListener("touchstart", this.handleInputBlur);
      this.drawerRef.addEventListener("resize", this.getDrawerStyle);
    }
  }

  componentDidUpdate() {
    this.init();
  }

  componentWillUnmount() {
    const darkOverlay = document.getElementsByClassName(
      this.darkOverlayClass
    )[0];
    if (this.drawerRef && isMobileWeb()) {
      this.drawerRef.removeEventListener("touchstart", this.handleInputBlur);
      this.drawerRef.removeEventListener("resize", this.getDrawerStyle);
    }
    if (darkOverlay) {
      darkOverlay.removeEventListener("click", this.closeNavOnOverlayClick);
    }
    enableBodyScroll(this.scrollTargetElement);
  }

  /**
   * To hide the keyboard in case user scrolls through navigation menu.
   */
  handleInputBlur = () => {
    if (document.activeElement.nodeName === "INPUT") {
      document.activeElement.blur();
    }
  };

  init = () => {
    const { open, renderOverlay } = this.props;
    if (!open) {
      removeBodyScrollLocks();
    }
    if (renderOverlay) {
      this.getDrawerStyle();
    }

    const darkOverlay = document.getElementsByClassName(
      this.darkOverlayClass
    )[0];
    if (open && darkOverlay) {
      darkOverlay.addEventListener("click", this.closeNavOnOverlayClick);
    } else if (darkOverlay) {
      darkOverlay.removeEventListener("click", this.closeNavOnOverlayClick);
    }
    return null;
  };

  /* Set drawer ref */
  setDrawerRef = (node) => {
    this.drawerRef = node;
  };

  /* Method to close nav bar on click of dark overlay */
  closeNavOnOverlayClick = () => {
    const { close, open } = this.props;

    if (open && typeof close === "function") {
      close();
      enableBodyScroll(this.scrollTargetElement);
    }
  };

  handleUserName = (userName) => {
    return userName.length <= 15
      ? userName
      : userName.substring(0, 15).concat("...");
  };

  handleUserRewards = (userRewards) => {
    return userRewards % 1 ? userRewards : Math.floor(userRewards);
  };

  disableScroll = (drawer) => {
    const { open } = this.props;
    this.scrollTargetElement = drawer;
    if (open && this.scrollTargetElement) {
      disableBodyScroll(this.scrollTargetElement);
    }
  };

  /**
   * Function to calculate the height of the visible smart banner for hamburger menu
   */
  getAppBannerHeight = () => {
    let bannerHeight = 0;
    const appBanner = document.getElementsByClassName("app-banner-wrapper")[0];
    const appBannerBoundaries = appBanner
      ? appBanner.getBoundingClientRect()
      : null;

    if (
      appBannerBoundaries &&
      appBannerBoundaries.top + appBannerBoundaries.height > -1
    ) {
      bannerHeight =
        appBannerBoundaries.top < 0
          ? appBannerBoundaries.height + appBannerBoundaries.top
          : appBannerBoundaries.height;
    }
    return bannerHeight;
  };

  /* Style for drawer to make it scrollable within */
  getDrawerStyle = () => {
    if (window) {
      const drawer = document.getElementById("tcp-nav-drawer");
      const headerTopNav = document.getElementsByClassName("header-topnav")[0];
      const middleNav = document.getElementsByClassName("header-middle-nav")[0];
      const condensedHeader = document.getElementById("condensedHeader");
      const userInfo = document.getElementById("sideNavUserInfo");
      const darkOverlay = document.getElementsByClassName(
        this.darkOverlayClass
      )[0];
      const userInfoHeight = userInfo
        ? userInfo.getBoundingClientRect().height
        : null;
      const wHeight = window.innerHeight;
      const {
        values: { lg },
      } = breakpoints;

      if (window.innerWidth < lg && drawer) {
        const headerTopNavComp = headerTopNav.getBoundingClientRect();
        const headerMiddleNavComp = middleNav.getBoundingClientRect();
        let headerHeight =
          headerTopNavComp.height +
          headerMiddleNavComp.height +
          this.getAppBannerHeight();

        if (headerTopNav && headerTopNavComp.top < 0) {
          headerHeight -= Math.abs(headerTopNavComp.top);
        }

        if (condensedHeader) {
          headerHeight = condensedHeader.getBoundingClientRect().height;
        }
        userInfo.style.top = `${headerHeight}px`;
        drawer.style.height = `${wHeight - (headerHeight + userInfoHeight)}px`;
        drawer.style.position = "fixed";
        drawer.style.overflowY = "scroll";
        drawer.style.top = `${headerHeight + userInfoHeight}px`;
        darkOverlay.style.top = `${headerHeight}px`;
        darkOverlay.style.position = "fixed";

        this.disableScroll(drawer);
      }
    }
  };

  render() {
    const {
      children,
      className,
      small,
      medium,
      large,
      open,
      id,
      close,
      renderOverlay,
      drawerFooter,
      hideNavigationFooter,
      showCondensedHeader,
      drawerElemId,
      navigationLevel,
    } = this.props;

    let openDrawer = open;
    if (typeof open === "string") {
      openDrawer = open === id;
    }
    // if (close && renderOverlay) {
    //   closeOverlay();
    // }
    // if (openDrawer && renderOverlay) {
    //   showOverlay();
    // }
    const classToOpen = openDrawer
      ? `tcp-drawer__isOpen_${navigationLevel}`
      : "";
    const condensedHeader = showCondensedHeader ? " tcp-condensed-drawer" : "";
    const classToHideOnViewports = hideOnViewport({ small, medium, large });
    const classToShowOnViewports = showOnViewport({ small, medium, large });

    return (
      <div className={className} ref={this.setDrawerRef} id={drawerElemId}>
        {
          // If Drawer is not required on all viewports then duplicate the DOM for the children without Drawer
          // User will have to handle display of this element with CSS
          isDrawerNotRequiredOnAllViewports(small, medium, large) && (
            <div className={`${classToShowOnViewports}`}>{children}</div>
          )
        }
        {openDrawer && (
          <React.Fragment>
            <aside
              className={`${classToOpen}${condensedHeader} ${classToHideOnViewports}`}
            >
              <div id="tcp-nav-drawer" className="tcp-drawer-content">
                {children}
              </div>
            </aside>
          </React.Fragment>
        )}
      </div>
    );
  }
}

export { Drawer as DrawerVanilla };
export default withStyles(Drawer, style);
