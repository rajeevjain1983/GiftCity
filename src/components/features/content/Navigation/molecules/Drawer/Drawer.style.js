// 9fbef606107a605d69c0edbcd8029e5d 
import { css } from 'styled-components';

export default css`
background:gray;
border:5px solid blue;


/* z-index: 1203; */

  /* @-webkit-keyframes slideIn {
    0% {
      transform: translateX(-900px);
    }
    100% {
      transform: translateX(0);
    }
  }
  @-moz-keyframes slideIn {
    0% {
      transform: translateX(-900px);
    }
    100% {
      transform: translateX(0);
    }
  }
  @keyframes slideIn {
    0% {
      transform: translateX(-900px);
    }
    100% {
      transform: translateX(0);
    }
  } */
  /* .tcp-drawer-overlay {
    background: ${props => props.theme.colors.common.black};
    position: absolute;
    z-index: 900;
    height: 100%;
    opacity: 0.6;
    width: 100%;
  }
  .is-hidden {
    display: none;
  }
  .tcp-drawer__isOpen_${props => props.navigationLevel} {
    display: block;
    position: absolute;
    z-index: 1200;
    top: ${props => (props.position && props.position.top) || 0};
    left: ${props => (props.position && props.position.left) || 0};
    width: 100%;
    height: ${props => props.height || `auto`};
  }
  .tcp-condensed-drawer {
    position: fixed;
    top: 58px;
  }
  .tcp-drawer-content {
    width: ${props => props.width.small || `100%`};
    background: green;
    height: ${props => props.height || `auto`};
  }
  .display-small-none {
    display: none;
  }
  
  .manage-account {
    position: absolute;
    top: 35%;
    right: 6%;
  }
  .rightLink {
    border-left: 1px solid ${props => props.theme.colors.gray[500]};
    box-sizing: border-box;
    margin-left: ${props => props.theme.spacing.ELEM_SPACING.XS};
    padding-left: ${props => props.theme.spacing.ELEM_SPACING.XS};
    min-width: ${props => props.theme.spacing.ELEM_SPACING.XXXL};
    border-radius: 0px;
  }
  .input-checkbox-wrapper {
    display: flex;
  }
  @media ${props => props.theme.mediaQuery.medium} {
    .tcp-drawer__isOpen {
      top: ${props => (props.position && props.position.topMedium) || 0};
    }
    .tcp-condensed-drawer {
      position: fixed;
      top: 62px;
    }
    .tcp-drawer-content {
      width: ${props => props.width.medium || `100%`};
    }
    .display-small-none {
      display: block;
    }
    .display-medium-none {
      display: none;
    }
  }
  @media ${props => props.theme.mediaQuery.large} {
    .tcp-drawer-content {
      width: ${props => props.width.large || `100%`};
      background: none;
    }
    .display-medium-none {
      display: block;
    }
    .display-large-none {
      display: none;
    }
  } */
`;
