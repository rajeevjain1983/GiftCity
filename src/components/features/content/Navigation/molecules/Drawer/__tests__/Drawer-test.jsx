// 9fbef606107a605d69c0edbcd8029e5d 
import React from 'react';
import { shallow } from 'enzyme';
import { DrawerVanilla as Drawer } from '../Drawer';

let DrawerComp;
beforeEach(() => {
  DrawerComp = shallow(
    <Drawer
      mobile
      tablet
      open
      navigationLevel="l1"
      width={{
        small: '314px',
        medium: '314px',
        large: '100%',
      }}
    >
      Renders correctly
    </Drawer>
  );
});

describe('Drawer component', () => {
  it('renders correctly', () => {
    expect(DrawerComp).toMatchSnapshot();
  });

  it('handleUserName should return the user name', () => {
    const instance = DrawerComp.instance();
    expect(instance.handleUserName('test')).toEqual('test');
  });
  it('handleUserRewards should get called', () => {
    const instance = DrawerComp.instance();
    expect(instance.handleUserRewards(1)).toEqual(1);
  });
  it('handleUserRewards should get called', () => {
    const shallowDrawerComp = shallow(
      <Drawer
        mobile
        tablet
        open={false}
        width={{
          small: '314px',
          medium: '314px',
          large: '100%',
        }}
      >
        Renders correctly
      </Drawer>
    );
    const instance = shallowDrawerComp.instance();
    expect(instance.init()).toBeNull();
  });
});
