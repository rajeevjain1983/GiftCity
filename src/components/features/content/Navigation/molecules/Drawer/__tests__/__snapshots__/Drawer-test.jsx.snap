// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer component renders correctly 1`] = `
<div
  id="drawer-wrapper"
>
  <div
    className="  "
  >
    Renders correctly
  </div>
  <aside
    className="tcp-drawer__isOpen_l1 display-small-none display-medium-none display-large-none"
  >
    <div
      className="tcp-drawer-content"
      id="tcp-nav-drawer"
    >
      Renders correctly
    </div>
  </aside>
</div>
`;
